org.jboss.tools.vscode.java/src/org/jboss/tools/vscode/java/internal/SharedASTProvider.java

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Red Hat Inc. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+package org.jboss.tools.vscode.java.internal;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.ISafeRunnable;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.SafeRunner;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ASTParser;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.internal.corext.dom.IASTSharedValues;
+
+/**
+ * AST provider that caches the AST generated for
+ * files that are known to be open.
+ *
+ * @author Gorkem Ercan
+ *
+ */
+@SuppressWarnings("restriction")
+public class SharedASTProvider {
+
+	private Map<String, CompilationUnit> cache = Collections.synchronizedMap(new HashMap<>());
+	private static SharedASTProvider instance;
+
+	private SharedASTProvider(){
+
+	}
+
+	public static SharedASTProvider getInstance(){
+		if(instance == null ){
+			instance = new SharedASTProvider();
+		}
+		return instance;
+	}
+
+	public CompilationUnit getAST(final ITypeRoot input,  IProgressMonitor progressMonitor) {
+
+		if (progressMonitor != null && progressMonitor.isCanceled())
+			return null;
+
+		final String identifier = input.getHandleIdentifier();
+		synchronized(cache){
```

Comment:
cache is already a synchronized map. Is it necessary to add an extra synchronized block?